src/test/app/Credentials_test.cpp

```diff
@@ -0,0 +1,943 @@
+        Account const issuer{"issuer"};
+        Account const subject{"subject"};
+        Account const other{"other"};
+
+        {
+            Env env{*this, features};
+
+            env.fund(XRP(5000), subject, issuer);
+
+            {
+                testcase("CredentialsAccept fail, Credential doesn't exist.");
+                env(credentials::accept(subject, issuer, credType),
+                    ter(tecNO_ENTRY));
+                env.close();
+            }
+
+            {
+                testcase("CredentialsAccept fail, invalid Issuer account.");
+                auto jv = credentials::accept(subject, issuer, credType);
+                jv[jss::Issuer] = to_string(xrpAccount());
+                env(jv, ter(temINVALID_ACCOUNT_ID));
+                env.close();
+            }
+        }
+
+        {
+            Env env{*this, features};
+
+            env.fund(drops(env.current()->fees().accountReserve(1)), issuer);
+            env.fund(drops(env.current()->fees().accountReserve(0)), subject);
+            env.close();
+
+            {
+                testcase("CredentialsAccept fail, not enough reserve.");
+                env(credentials::create(subject, issuer, credType));
+                env.close();
+
+                env(credentials::accept(subject, issuer, credType),
+                    ter(tecINSUFFICIENT_RESERVE));
+                env.close();
+            }
+        }
+
+        {
+            using namespace jtx;
+            Env env{*this, features};
+
+            env.fund(XRP(5000), subject, issuer);
+            env.close();
+
+            {
+                env(credentials::create(subject, issuer, credType));
+                env.close();
+
+                testcase("CredentialsAccept fail, invalid fee.");
+                auto jv = credentials::accept(subject, issuer, credType);
+                jv[jss::Fee] = -1;
+                env(jv, ter(temBAD_FEE));
+
+                testcase("CredentialsAccept fail, lsfAccepted already set.");
+                env(credentials::accept(subject, issuer, credType));
+                env.close();
+                env(credentials::accept(subject, issuer, credType),
+                    ter(tecDUPLICATE));
+                env.close();
+            }
+
+            {
+                const char credType2[] = "efghi";
+
+                testcase("CredentialsAccept fail, expired credentials.");
+                auto jv = credentials::create(subject, issuer, credType2);
+                uint32_t const t = env.now().time_since_epoch().count();
+                jv[sfExpiration.jsonName] = t;
+                env(jv);
+                env.close();
+
+                // credentials are expired now
+                env(credentials::accept(subject, issuer, credType2),
+                    ter(tecEXPIRED));
+                env.close();
+
+                // check that expired credentials were deleted
+                auto const jDelCred =
```

Comment:
Please also check the owner directories (should just be a matter of checking the `account_objects` length).

Reply:
I don't think it is needed here as this test doesn't check for inconsistency among ledger_entry / account_objects / ledger_data ... etc.  
`ledger_entry`  doesn't take any "**root_account**" parameters, so we should trust it that object doesn't exists in whole ledger.

Reply:
It needs to correctly be removed from the owner directories, though. I'd like to have a test that checks for that.

Reply:
How it is possible that the reference to the object will be present in one of the directories and `ledger_entry` will not show it?

Reply:
At the least, the owner counts should be checked, since that is a separate line of code.

Reply:
Added ownerCounts to credentials and depositPreauth tests